platform-sdk/docs/proposals/TSS-Ledger-Id/TSS-Ledger-Id.md

```diff
@@ -0,0 +1,826 @@
+
+#### Ledger Signature API
+
+The Ledger Signature API is the interface between the application and the platform by which the application submits
+a message to be signed by the ledger, the nodes sign the message with their shares, and the platform aggregates the
+signatures into a ledger signature. The ledger signature is then asynchronously returned to the application.
+
+To request a ledger signature, the application invokes a new method on the `State`:
+
+* `State::sign(byte[] message) throws NotReadyException`
+
+To receive a ledger signature, the application registers a callback with the platform builder:
+
+* `PlatformBuilder::withLedgerSignatureCallback(Consumer<PairingSignature> ledgerSignatureConsumer)`
+
+#### System Transactions
+
+The following are new system transactions needed to key rosters and generate ledger signatures.
+
+##### TssMessage System Transaction
+
+TODO: Protobuf of data
+
+##### TssVote System Transaction
+
+TODO: Protobuf of data
+
+##### TssShareSignature System Transaction
+
+TODO: Protobuf of data
+
+#### State Datastructures
+
+The state needs to store the relevant ledger id, consensus rosters, key material, and partially constructed ledger
+signatures.
+
+##### Ledger Id
+
+The `ledgerId` is the public ledger key able to verify ledger signatures. It is used by both the application and the
+platform. This value should not change during address book changes and its value does not change unless the network
+goes through a another TSS bootstrap process. Since its value is not expected to change, storing it in a singleton
+Merkle Leaf by itself is appropriate.
+
+TODO: Protobuf of data
+
+##### Consensus Rosters
+
+This proposal extends the roster data format where each `RosterEntry` has a new long term public EC key called a
+`tssEcKey` that is used in the Groth21 algorithm.
+
+TODO: Protobuf of data
+
+##### Key Material (TSS Data Map)
+
+The TSS Data Map is a Map Merkle Leaf that is a combined map of all the key material for all rosters that are
+tracked and the votes by nodes that have validated the key material.
+
+Keys in the TSS Data Map have the following structure: (KeyType, RosterHash, SequenceNumber).
+
+1. KeyType - An enum that is one of TSS_MESSAGE or TSS_VOTE.
+2. RosterHash - The hash of the consensus roster that the message or vote is related to.
+3. SequenceNumber - The order in which the message or vote came to consensus.
+
+The value associated with a TSS_MESSAGE key is the pair (ShareId, TssMessage)
+
+1. ShareId - The id of the share that the message is related to.
+2. TssMessage - the raw bytes of the TssMessage
+
+The value associated with a TSS_VOTE key is a bit vector with the following interpretation
+
+* The order of bits from the least significant bit to the most significant bit corresponds to the numeric order of
+  share ids from least to greatest.
+* If a bit is set to 1, then the TssMessage for the corresponding ShareId was valid and contributed to a successful
```

Comment:
> and contributed to a successful reconstruction of the ledger id.

I think I understand what you are getting at, but this isn't fully correct. Take, for example, the case where the threshold of valid messages was NOT met. In this case, a node should vote `1` for the messages that were valid, even though these messages did not contribute to a successful reconstruction.

Maybe the way to express this idea is to say that valid messages receive a vote of `1`, but that nodes only vote `1` for the first `threshold` number of valid messages (in consensus order), and that all subsequent messages receive a vote of `0`

Reply:
I think I have fixed the wording.  Even when the threshold is not met, the tss message would only get a 1 if it is one of the first `threshold` number of valid TssMessages in consensus order.